### ISSUE TITLE
Cannot archive with Xcode 11 GM 2

### ISSUE THREAD

**Issue by WillBishop** [scinfu/SwiftSoup#125]
Hello,

I'm unable to archive SwiftSoup with Xcode 11 GM 2. it hangs during the Xcode archive process (seemingly on a different framework, but removing SwiftSoup fixes it)

I've seen issues #116, but it was closed without a resolution for me.

I've tried @Dschee's and @kkla320's forks/PRs, but neither work for me.

How could I get some info that would help you debug this?

**Comment by scinfu** [comment]
can you provide a demo project with this problem?
Thank you

**Comment by lucascorrea** [comment]
> Hello,
> 
> I'm unable to archive SwiftSoup with Xcode 11 GM 2. it hangs during the Xcode archive process (seemingly on a different framework, but removing SwiftSoup fixes it)
> 
> I've seen issues #116, but it was closed without a resolution for me.
> 
> I've tried @Dschee's and @kkla320's forks/PRs, but neither work for me.
> 
> How could I get some info that would help you debug this?

Same problem, Xcode stop building on SwiftSoup

**Comment by lucascorrea** [comment]
`/Users/lucas/Library/Developer/Xcode/DerivedData/LeapcardViewer-fgxzikddiauoyufgowiwsldcsser/Build/Intermediates.noindex/ArchiveIntermediates/LeapcardViewer/IntermediateBuildFilesPath/Pods.build/Release-iphoneos/SwiftSoup.build/Objects-normal/armv7/XmlTreeBuilder.bc`

I've tried to build by command line always stop in that line.

**Comment by scinfu** [comment]
@lucascorrea can you provide a demo project with this problem?

**Comment by cristosv** [comment]
Same issue with Xcode 11 GM2.  SwiftSoup does compile on an empty iOS 13 template project, but for some reason, hangs when compiled within an app upgraded from XCode10/Swift 4.2 to XCode11/Swift 5.0.   I have tried installing as a cocoapod and as a swift package.  Using cocoapods and changing the build optimization of SwiftSoup to -Onone compiles/archives.

**Comment by scinfu** [comment]
Hi
i added new version 2.2.1.
Previously made fixes were not included in a version.
Let me know if the problem persists

**Comment by CentrumGuy** [comment]
> Hi
> i added new version 2.2.1.
> Previously made fixes were not included in a version.
> Let me know if the problem persists

Pod update doesn't update SwiftSoup to version 2.2.1. It thinks that the newest version is version 2.2.0

**Comment by phr85** [comment]
To ensure I use the latest version:
pod 'SwiftSoup', :git => 'https://github.com/scinfu/SwiftSoup.git', :commit => 'b9f0389'

But I still can't archive my App with Xcode 11. But not sure if there is a other framework who makes trouble.

**Comment by scinfu** [comment]
@phr85 i pusched 2.2.1 alto for cocoapods.
i use pod 1.7.5 and in cocoapods there is no problem.

Someone can provide a test project with this bug? Please.
Thank you

**Comment by CentrumGuy** [comment]
Updated to 2.2.1 but I'm still experiencing the issue

**Comment by scinfu** [comment]
Someone can provide a test project with this bug? Please.
Thank you

**Comment by CentrumGuy** [comment]
Here you go (this is the FirebaseFirestore project I created for [https://github.com/firebase/firebase-ios-sdk/issues/3884](url) so you'll have to create a corresponding Firebase project on the Firebase website for this to run correctly but it still shouldn't archive either way): https://github.com/CentrumGuy/Xcode-Archive-Bug.git

**Comment by CentrumGuy** [comment]
Here you go (this is the FirebaseFirestore project I created for [https://github.com/firebase/firebase-ios-sdk/issues/3884](url) so you'll have to create a corresponding Firebase project on the Firebase website for this to run correctly but it still shouldn't archive either way): https://github.com/CentrumGuy/Xcode-Archive-Bug.git

**Comment by phr85** [comment]
I can confirm:
I removed SwiftSoup 2.2.1. Now the project Archiving as it should. 
I was now about to make a sample project, but @CentrumGuy was faster :-)

**Comment by noahemmet** [comment]
Here's an open source project where I'm running into the same problem, in case that's useful: https://github.com/noahemmet/calagator

**Comment by sergstav** [comment]
2.2.1 dont archive.

**Comment by noahemmet** [comment]
Just want to note that because [my project](https://github.com/scinfu/SwiftSoup/issues/125#issuecomment-533850496) uses Swift Package Manager, the .xcodeproj's build settings won't have any effect on the final dependency. Thus, I don't think the optimization settings are at fault here. Is this issue possibly related to [New swift-corelibs-foundation module: FoundationXML](https://forums.swift.org/t/new-swift-corelibs-foundation-module-foundationxml/27544)?

**Comment by kreeger** [comment]
I was only able to get my project to build with the latest version of SwiftSoup by disabling Swift compiler optimization. This is naturally sub-optimal, but this was the only CocoaPod I had to do this to.

```ruby
post_install do |installer|
  installer.pods_project.targets.each do |target|
    next unless target.name == 'SwiftSoup'
    target.build_configurations.each do |config|
      next unless config.name.start_with?('Distribution')
      config.build_settings['SWIFT_OPTIMIZATION_LEVEL'] = '-Onone'
    end
  end
end
```

**Comment by kreeger** [comment]
Here we go! A fresh project, created in Xcode 11, with SwiftSoup added via SwiftPM. That's all that's required to reproduce this issue.

https://github.com/kreeger/SwiftSoupTest

**Comment by noahemmet** [comment]
And just to clarify—this happens for any `RELEASE` build, not just during archive.

**Comment by kreeger** [comment]
I should also say, I upgraded from v1.7.4 hoping that v2.2.1 would fix the issue, before finding this thread — so whatever's funky _might_ have been a problem in SwiftSoup for a while.

**Comment by FrankB1708** [comment]
Man, I’ve been struggling with this for weeks trying to publish an update to my app with Xcode 11.
Removing SwiftSoup from my project or compiling it in debug mode does fix the issue for me.

I honestly believe this is an Xcode bug: it should not be just hanging on a file forever with no error. That being said, given how slow is Apple to correct Xcode bugs, it would be much better for SwiftSoup to find a workaround for this XCode bug.

**Comment by WillBishop** [comment]
It’s probably both a SwiftSoup bug and a Swift Toolchain bug.

A compiler hang is *always* a compiler bug, it should tell you what it got hung on. But that said, it’s probably hung on some SwiftSoup setting or file

**Comment by nk1902** [comment]
Had no issues archiving on Xcode 10.2, but with Xcode 11 it's hanging... hm

**Comment by nk1902** [comment]
> I was only able to get my project to build with the latest version of SwiftSoup by disabling Swift compiler optimization. This is naturally sub-optimal, but this was the only CocoaPod I had to do this to. (I know this has been discussed in other threads, too.)
> 
> ```ruby
> post_install do |installer|
>   installer.pods_project.targets.each do |target|
>     next unless target.name == 'SwiftSoup'
>     target.build_configurations.each do |config|
>       next unless config.name.start_with?('Distribution')
>       config.build_settings['SWIFT_OPTIMIZATION_LEVEL'] = '-Onone'
>     end
>   end
> end
> ```
> 
> I'm going to see if I can come up with an example project with _just_ `SwiftSoup` included.

where do you put this block to get the build to archive?

**Comment by jatyson** [comment]
@nk1902 Put it at the bottom of your Podfile.

**Comment by gbuela** [comment]
> I was only able to get my project to build with the latest version of SwiftSoup by disabling Swift compiler optimization. This is naturally sub-optimal, but this was the only CocoaPod I had to do this to. (I know this has been discussed in other threads, too.)
> 
> ```ruby
> post_install do |installer|
>   installer.pods_project.targets.each do |target|
>     next unless target.name == 'SwiftSoup'
>     target.build_configurations.each do |config|
>       next unless config.name.start_with?('Distribution')
>       config.build_settings['SWIFT_OPTIMIZATION_LEVEL'] = '-Onone'
>     end
>   end
> end
> ```
> 
> I'm going to see if I can come up with an example project with _just_ `SwiftSoup` included.

Thanks. I thought this wasn't working for me until I realized I have no configurations named Distribution*, just vanilla Debug and Release. 

Hopefully the actual issue gets fixed soon...

**Comment by funkenstrahlen** [comment]
Has anyone tried this with Xcode 11.1 GM released yesterday?

**Comment by noahemmet** [comment]
@funkenstrahlen Good question! I just tried it on the GM and it still hangs.

**Comment by sphilipakis** [comment]
> Has anyone tried this with Xcode 11.1 GM released yesterday?

yes, it still hangs on Xcode 11.1 GM

**Comment by recre4te** [comment]
I can't archive the project with SwiftSoup (2.2.1)

**Comment by phr85** [comment]
> I can't archive the project with SwiftSoup (2.2.1)
@recre4te: That's exactly the issue we discuss here ;-) See the temporary solution of @kreeger:
Add this in your Podfile at the end of the file:

```
post_install do |installer|
  installer.pods_project.targets.each do |target|
    next unless target.name == 'SwiftSoup'
    target.build_configurations.each do |config|
      next unless config.name.start_with?('Distribution')
      config.build_settings['SWIFT_OPTIMIZATION_LEVEL'] = '-Onone'
    end
  end
end
```

**Comment by GrLb71** [comment]
> @recre4te: That's exactly the issue we discuss here ;-) See the temporary solution of @kreeger:
> Add this in your Podfile at the end of the file:

Is there any temporary solution when using Swift Package Manager?

**Comment by funkenstrahlen** [comment]
> Is there any temporary solution when using Swift Package Manager?

Unfortunately not. There is now way to adjust build settings.

**Comment by L4grange** [comment]
Thank you for the workaround, it was driving mad that Xcode would not finish archiving when I need to ship my iOS 13 fixes. In my case I had to replace `config.name.start_with?('Distribution')` with `config.name.start_with?('Release')` since I haven't made any changes to the default Xcode project, as @gbuela suggests. Hope this gets sorted out soon enough (especially on the Apple side, hanging the build with no error message or timeout is a mess)

**Comment by damienissa** [comment]
Hi, guys. I have simple solution. Just use Carthage.

**Comment by CentrumGuy** [comment]
> Hi, guys. I have simple solution. Just use Carthage.

I don't think that that's a good solution since my project (and I assume almost everyone else's) is already configured with cocoapods and it would be a difficult process to convert from the cocoapods configuration to Carthage.

**Comment by jdanthinne** [comment]
This is my version of @kreeger solution. Had to change `Distribution` to `Release`, but also `target.name == 'SwiftSoup'` to `target.name.start_with?('SwiftSoup')` (because the target name is "SwiftSoup-iOS" or "SwiftSoup-macOS").

```Ruby
post_install do |installer|
  installer.pods_project.targets.each do |target|
    next unless target.name.start_with?('SwiftSoup')
    target.build_configurations.each do |config|
      next unless config.name.start_with?('Release')
      config.build_settings['SWIFT_OPTIMIZATION_LEVEL'] = '-Onone'
    end
  end
end
```

**Comment by 224XS** [comment]
Sadly the conditional configuration from @kreeger and modded by @jdanthinne does not work in my project.  I am trying to update an app to iOS 13 features and when I archive to a Release config, Xcode just
hangs in the Build target SwiftSoup just after the "touch".  Im using the latest release Xcode Version 11.0 (11A420a).

I also trying changing the  Swift Optim. build settings directly in the project, again no joy.

**Comment by noahemmet** [comment]
(Still a problem with SPM and Xcode 11.2 beta 2.)

**Comment by noahemmet** [comment]
Also, I don't know if it's been mentioned, but the SwiftSoup project *itself* won't compile as `RELEASE` when opened via `Package.swift`. In other words, this isn't only a problem when SwiftSoup is imported as a dependency.

I've been going through the files and replacing all function bodies with `fatalError()`s, just to see if there's some sort of compiler type-checker bug, but no luck so far.

**Comment by sheltonsbs** [comment]
> Sadly the conditional configuration from @kreeger and modded by @jdanthinne does not work in my project. I am trying to update an app to iOS 13 features and when I archive to a Release config, Xcode just
> hangs in the Build target SwiftSoup just after the "touch". Im using the latest release Xcode Version 11.0 (11A420a). Using SwiftSoup 2.2.1
> 
> I also trying changing the Swift Optim. build settings directly in the project, again no joy.

I tried the custom config in podfile and failed the first time too.
'''
      if target.name == 'SwiftSoup'
        config.build_settings['SWIFT_OPTIMIZATION_LEVEL'] = '-Onone'
      end
'''

Then I did a '''pod install''' on terminal 
then clean using '''xcodebuild clean -workspace [workspaceName] -scheme [schemaName]'''
And of course archive ''''xcodebuild clean -workspace [workspaceName] -scheme [schemaName]'''
(since xcode is crashing on pod install, I've been trying to build via terminal. Xcode should work too)

Hope these steps can help :)

**Comment by funkenstrahlen** [comment]
Issue not fixed in Version 11.2 beta (11B41).

**Comment by furiosFast** [comment]
same problem for me on xcode 11

**Comment by kiliankoe** [comment]
> config.build_settings['SWIFT_OPTIMIZATION_LEVEL'] = '-Onone'

In case it helps finding the issue, running `swift build -Xswiftc -O` in SwiftSoup leads to the compiler apparently getting stuck on Node.swift.

**Comment by iva1ex** [comment]
Same problem for me.
Archiving with SwiftSoup hangs on Xcode 11. 
Also, building SwiftSoup via Terminal command `xcodebuild` hangs on compiling `XmlTreeBuilder.swift` file.

**Comment by kiliankoe** [comment]
This appears to be fixed on Swift master 👍 

https://bugs.swift.org/browse/SR-11577?focusedCommentId=50678&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-50678

**Comment by tylerdak** [comment]
When, would you say, will we get to see that fix, then?

**Comment by sergstav** [comment]
It is ok now for Xcode 11.1? Archiving oK?

**Comment by GrLb71** [comment]
> It is ok now for Xcode 11.1? Archiving oK?

No. You need new toolchain compiler as metioned above. Xcode 11.1 is not enaugh.

**Comment by 224XS** [comment]
The new compiler/tool chain Does compile/run/archive just fine!

However, archive compile on non-canonical tool chains are NOT uploadable to the App Store!

We just have to wait for Apple to approve and distribute a new update for Xcode with the new tool chain.

**Comment by tylerdak** [comment]
This is working for me:

Add the following to the bottom of the Podfile that should contain SwiftSoup:
`post_install do |installer|
  installer.pods_project.targets.each do |target|
    next unless target.name == 'SwiftSoup'
    target.build_configurations.each do |config|
      next unless config.name.start_with?('Release')
      config.build_settings['SWIFT_OPTIMIZATION_LEVEL'] = '-Onone'
    end
  end
end`

Then, do `pod install`.
Try archiving with the default toolchain now.

**Comment by nirajjoshi** [comment]
> This is working for me:
> 
> Add the following to the bottom of the Podfile that should contain SwiftSoup:
> `post_install do |installer|
> 
> installer.pods_project.targets.each do |target|
> 
> ```
> next unless target.name == 'SwiftSoup'
> 
> target.build_configurations.each do |config|
> 
>   next unless config.name.start_with?('Release')
> 
>   config.build_settings['SWIFT_OPTIMIZATION_LEVEL'] = '-Onone'
> 
> end
> ```
> 
> end
> 
> end`
> 
> Then, do `pod install`.
> Try archiving with the default toolchain now.

Is it also uploadable to the App Store?

**Comment by tylerdak** [comment]
I'm able to get it to TestFlight, so yeah I think so.

**Comment by 224XS** [comment]
For at least some of us using SwiftSoup, the various switch setting for optimization did nothing (see above). The actual fix is a change in the compiler logic having to do with an infinite loop potential in handling case statements. I’m going to wait for Apple to push the new compiler.

**Comment by tylerdak** [comment]
Oh, I'm sorry, not sure I saw that above. My bad.

**Comment by deviosap** [comment]
<img width="159" alt="Screenshot 2019-10-12 at 16 08 05" src="https://user-images.githubusercontent.com/49749684/66702782-afb7eb00-ed0b-11e9-84a3-1e381f5ee45a.png">
<img width="561" alt="Screenshot 2019-10-12 at 16 10 32" src="https://user-images.githubusercontent.com/49749684/66702785-b6466280-ed0b-11e9-93e1-5b7fb57906d8.png">

I have found a clean and quick solution in the meantime, click "Pods" in project navigator, then in targets (picture) set SwiftSoup,enter image description here in Swift compiler set both -Onone. Now try to archive. enter image description here

**Comment by sergstav** [comment]
So, changing all optimization level for SwiftSoup to None works for me, and also normal upload to Testflight.

![Screenshot 2019-10-15 at 14 46 30](https://user-images.githubusercontent.com/3393265/66831794-2ac60f00-ef61-11e9-88ec-6cdf5f3825ad.png)

**Comment by 224XS** [comment]
Update Report:
After sucessfully playing with the revised compiler for a few days, I went back to the release version toolchain in Xcode 11, shut off the Swift Code Gen optimization as before, and the Clang Code Gen Optimization too this time, and lo and behold, the Archive sailed right into the Store!

**Comment by nostradani** [comment]
I stumbled over this issue today, after we upgraded Xcode to version 11.1

We encountered similar issues with other projects in the past where the swift compiler would crash due to some code it couldn't handle properly. 

In order to find the responsible line of code, I did the following:
- Change the build setting for SwiftSoup from "Whole Module" to "Incremental"
- In the build log I could now see the swift files where it got stuck ("Though not all the files were an issue")
- Then I commented out as much logic as possible in the files until the build went through.
- Uncommenting single lines and building again would then reveal the lines responsible for the issue.
- Refactor the responsible code lines until it builds.

I created a pull request with the change:
https://github.com/scinfu/SwiftSoup/pull/133

With this my project and also a sample project provided in this issue tread will now build in Xcode 11.1

Please note that this change merely avoids the hang issue.The real issue lies with the swift compiler since it should either be able to compile or at least provide an error. However I'm not familiar on how to best report this issue with the swift open source project.

**Comment by BreeeZe** [comment]
XCode 11.2 was just released and it is able to successfully archive with the current SwiftSoup release!

**Comment by funkenstrahlen** [comment]
Can confirm it works fine with Xcode 11.2. finally.

**Comment by iva1ex** [comment]
The hanging issue was fixed in this PR:
https://github.com/scinfu/SwiftSoup/pull/133

The new version v2.2.2 was released with a fix! 🎊

**Comment by BreeeZe** [comment]
Well yes, that fixes it too, but Xcode 11.2 now also works with v2.2.1 😄
